DM-28737 (#1)
lsst-dm/dmtn-183: index.rst

:tocdepth: 1

.. sectnum::

.. note::

   **This technote is not yet published.**

   Design document for a database of all alerts sent by the Rubin observatory.

Summary
=======

This document proposes a technical design for a database of alert packets.

At a high level, the proposal is to store raw alert packets in an object store.
All packets can be retrieved by ID.
Queries which search the alert database can be performed by querying the PPDB to get IDs of alert packets, and then requesting the packets by ID from the object store.

> In addition, to facilitate bulk access, we propose storing single-file archives of all alerts published each night.

Reviewer: ```suggestion
In addition, to facilitate bulk access, we suggest storing single-file archives of all alerts published each night.  
```

Despite the ZTF experience I'm really not sure if the demand warrants this additional product?  It seems expensive from a storage point of view.  Are you thinking that this would relieve some of the pressure on the object store?

Author: Yes; I interpret some of the use cases as pushing us in this direction for bulk export so people can train algorithms against large swaths of the data.

I'm always happy to shrink scope, though.

Reviewer: Let's leave it open as a possibility and talk with users & management.

Author: Made https://jira.lsstcorp.org/browse/DM-30016, will update the text

This system should be straightforward to build and administer, and it should be cost effective.

> It will require about 2-3 terabytes of space in the object store per year, and about 40 gigabytes of space per year for the PostgreSQL index database.

Reviewer: Is the "PostgreSQL index database" separate from the PPDB?  It's not discussed further in the document that I see.

Author: Good catch - this is a holdover from an earlier revision. There's no index, you're right.

Use cases: queryable DB
^^^^^^^^^^^^^^^^^^^^^^^

The Alert Database needs to support "needle in a haystack" queries based on specific a single alert identifier.
In particular, `get(alert_id)`` ought to retrieve the single alert packet associated with a specific alert_id.

This should generally respond within a few seconds.

In general, we can expect recent data to be queried more often than old data for this purpose, but everything needs to be available.

Use cases: resource for developers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Users who want to test algorithms, generate simulated streams, or evaluate filters will be interested in retrieving large batches of alert packets.
They will not need to retrieve these batches very often, and it's entirely acceptable for their requests to take a long time (even days) to fulfill.

Data Sizing Calculations
------------------------

DMTN-102 :cite:`DMTN-102` provides estimates for the total volume of alerts received.
We expect a maximum of about 10 million alerts per night, each about 80 kilobytes.
LSE-163 :cite:`LSE-163` projects about 300 nights of observations per year.

> Combining these figures, we project 80 gigabytes of alert data per night, and 2.4 terabytes per year.

Author: This math is wrong by a factor of 10: 80KB * 10 million alerts = 800 gigabytes.

cc @ktlim and @ebellm

Reviewer 2: Doh.  Can't believe I missed that.  Also, 800*300 = 240 TB, no?  That starts to make space-saving look better.  What about compression?

Author: Gah, yes, more factors! Thanks, yes, that's significant.

Author: > What about compression?

So, Avro files support a bunch of compression schemes: deflate, gzip, zstd, snappy, and bz2. By "support" I mean that they store the name of the compression scheme in a plaintext header, and then the data portion of an Avro file is the chunk that's compressed.

I haven't measured compression ratios on LSST alert packets. It's worth testing, based on this volume of data.

One option to boost compression ratios might be to try to reuse dictionaries across multiple alert packets. We could do that by storing many alert packets in one file, or by being very clever and careful with how we initialize our Avro file writers to reuse a compressor. But first - I should figure out whether naive compression does a good job.

---

Also, Avro Object Container format would mean including a 50KB schema in the plaintext header. That's not compressed by Avro. If we used one alert per file, then we'd have one schema per alert. Storing 50KB * 10 million alerts * 300 nights = 150 TB per year, which is just the same JSON document repeated 3 billion times, does not seem wise.

Author: I measured compression of the [sample batch of alerts based on HITS](https://github.com/lsst-dm/sample_alert_info/):

| codec | file with 1 alert | file with 50 alerts | file with 217 alerts |
| --------- | --------------------| ------------------------|----------------- |
| null | 93 KB | 2.352 MB | 22.426 MB |
| deflate | 67 KB | 1.113 MB | 15.454 MB |
| snappy | 71 KB | 1.250 MB | 17.592 MB |
| ztsd | 68 KB | 1.113 MB | 15.464 MB |
| bz2 | 69 KB | 1.117 MB | 15.583 MB |

(the file with 217 alerts may compress worse because it includes some alerts with very large cutouts, which probably compress poorly)

So we're basically looking at a 20-60% reduction in volume by using compression, something in that range.

Author: I have made some major revisions to try to deal with this:
 - gzip every alert
 - store schemas separately from alert packet data
 - drop the idea of storing a duplicate copy for bulk export - bulk export will just be slow, thats OK

 - DMS-PRTL-REQ-0048 "Alert Visualization" :cite:`LDM-554`:

     The Portal aspect shall provide for the users a "property sheet" for the contents of an alert packet including, but not necessarily limited to, the alert postage stamp image, the postage stamp time series, the photometric time series, the source and object information (e.g., position, brightness).

Proposed Implementation
=======================

We can satisfy these design inputs by storing serialized Avro alert data (the same bytes sent via Kafka to brokers) in a S3-like object store, indexed by a unique alert ID.
Each alert packet corresponds to one object in the object store.

.. note::

   An alternative would be to combine many packets into a block in the object store, perhaps of about 100 alert packets.
   This might permit more efficient storage and retrieval.
   Storage might be more efficient because compression would be better when storing many alerts.
   Retrieval might be more efficient because it might tune the outgoing flows into a smaller number of TCP connections which get a chance to grow window sizes.

>    But this would be more complex, and make writing more difficult, as writes need to append to existing data which would require coordination between writers, so this design sticks to a simpler structure.

Reviewer: It's worth considering that there is a natural unit of storage, which is the CCD-visit.  Each alert generation process will only be handling a single CCD of data, so all of the alerts from that CCD could be batched and stored together.  I do not know at the moment if that would make a substantial difference in latency--I do not believe there is much parallelism within the relevant tasks at present, though.

Reviewer: Another question raised by the discussion on DM-28734 is whether the alerts are stored on disk with schemas.  We are not transmitting alerts with schemas for bandwidth reasons, so a verbatim archive of what was transmitted would not include the schemas.  However, if we archive the alerts with the schemas users (and our own services) will not need to retrieve the appropriate schemas from elsewhere.

Since Avro schema storage is quite inefficient, archiving alerts with schemas might push us somewhat in the direction of a single file (with schema) per CCD.  

If we did store multiple alerts as a single object, in order to retrieve a single alert the alertDB service would have to seek through the file to find it, which would add latency.   That could hinder using the alertDB as a key component of a hybrid alert system.

Author: Interesting point, yeah.

The space savings probably approaches 1.5x to 2x, which doesn't seem like enough to really be a decider.

A more compelling reason to do this, though, would be that it reduces CPU burden on callers. Parsing the entire schema document for _every_ file is very wasteful; parsing schemas is usually not very well-optimized in Avro libraries since it's expected that it's done just once. We could remediate that with a bit of extra complexity: we could run a fronting service which strips schemas from alerts when they're requested.

As you say, an alertDB service would need to seek for a specific alert in a file to respond with it. But how much latency would this add? 

A single CCD-visit is expected to have about `10,000 / 189 = 52` alerts. How long does it take to scan 52 alerts?

I don't know exactly, but I can estimate based on [Avro benchmarks of ZTF data](https://github.com/fastavro/fastavro/issues/503) run on my laptop. A naive decoder with the `fastavro` library takes about 1.5ms per alert; a tuned hand-written one takes about 1ms per alert. So we're looking at adding something like 50ms to 70ms - which is actually quite significant, much more than I would have ever expected.

Reviewer 2: Keep in mind that it is possible for the alert schema to evolve over time, which might affect how schemas are stored and transmitted.

Author: Right, schema evolution is a very good reason to store schemas directly in files. It's actually [mandated in the Avro specification for that reason](http://avro.apache.org/docs/current/spec.html#Data+Serialization+and+Deserialization):

>  files or systems that store Avro data should always include the writer's schema for that data.

To my mind, there's no real question that we should include the schema in the files we store. But we could amortize the overhead of that inclusion by storing many alerts in a single file, like Eric says. I think we'd need to record byte offsets of each sub-alert in order to make that efficient though.

---

One unorthodox way we could do that would be to include a mapping of `alertID -> integer index of the alert within the file`. This could be included as a `map`-typed value in the metadata header of an Avro Object Container file.

Next, each alert could be stored a single "data block" within the Avro Object Container file. Data blocks are always prefixed with their size in bytes, which lets readers efficiently skip entire blocks.

Then, the seeking will be pretty close to free. We read (and parse) the Avro schema which is embeded in the file, and then read the header metadata to figure out how many blocks to skip. We skip each block based on the byte size field, and then actually extract the alert packet only when we have hit the block we're seeking for.

This is fairly complex stuff and uses parts of Avro that are rarely exposed in Python libraries, but I don't think it's very difficult really, and the complexity could be well-encapsulated, either in a Python module that people use (an AlertDB API client, I guess), or a remote service that does this on their behalf.

Author: What do you think about the above? The tradeoff is essentially:

 - More code complexity
 - Less disk usage in the object store (The 50KB schema is stored once per ~50 images instead of once per image, so saves about 50KB per alert, which is about ~20TB~ 150TB per year.

Bandwidth savings are a wash: we would save sending the schema, but need to send a bunch of superfluous alerts when plucking a single one out, even if we skip them.

CPU savings are probably a wash too.

Reviewer 2: If we're using a custom reader that is not simply returning the object in the object store (so that there's no need for that object to conform to Avro specifications), then I think it's far simpler to store the schema in a detached form (once per schema evolution change) and attach it to the object upon reading.

But it would be nice to not need the custom reader (which then requires the schema to be attached directly).  Across-object compression isn't a thing, I guess...

Author: I've revised to say that we should store schemas separately, using a 4-byte ID.

> An object store is used because it is cheap, scales well to handle terabytes of data, and should support parallel retrieval reasonably well.

Reviewer: I'd suggest saying a little more about what technical features the object store needs or doesn't.

Author: I haven't forgotten this - I can get to it when we feel confident that we have the rest of the doc roughly right.

Writing data
------------

> When the alert production pipeline has computed a new alert packet, it should be careful to write it to the alert archive before publishing to a Kafka topic, to ensure that it gets archived. The overall flow should be:

Reviewer: While I agree that this is necessary to ensure data is not lost, it creates additional latency in releasing the alerts, and the 60 second alert latency requirement is high priority.  We would need to ensure ~sub-second writes to the alert db or consider alternative approaches.

Reviewer: If that latency to the archive is too hard to guarantee, I wonder if temporary staging to a local filesystem would be sufficient.

Author: ~sub-second median write times is almost certainly easy. ~sub-second p99.9 write times is almost certainly extremely difficult and would require careful TCP tuning. (The same is true of the writes to Kafka, by the way).

I don't think hard time limits are compatible with hard data durability requirements. There's probably a formal way to state that (maybe [PACELC?](https://brooker.co.za/blog/2015/09/26/cap-durability.html)). I think we have to choose one or the other.

Temporary staging to a local filesystem doesn't seem to guarantee durability to me; I think it pushes the problem into an additional system (if you write to Kafka after the filesystem write, then you're now depending upon the filesystem -> DB flow working flawlessly, which is not realistic).

Reviewer 2: I'd be OK with us having an agreement with downstream brokers to recover alert packets from them in case we had an unexpected inability to write to the AlertDB and couldn't recover from our own queues.  This would mean that the AlertDB might not always have an alert available at the same time that a hybrid alert mini-packet went out, but I think downstream services need to protect against service-level unavailability anyway, so alert-level unavailability is a specific case of the general problem.

Author: @ktlim's idea of recovering alerts from downstream brokers is good, and it suggests a further change. We could write to AlertDB from a consumer hanging off of the alerts Kafka topic.

This means we are writing to the alert database after we've published alerts, but Kafka could keep a healthy buffer (7 days, or something) so we can be confident _enough_ that we'd preserve the alert in a database.

This would add latency to the writes to the Alert DB, but that seems acceptable to me - very low latency is not one of the driving goals of the thing.

Reviewer: I'm flexible here; as I said, keeping latency low is important.  Using Kafka as an intermediate store is attractive since brokers don't get alerts any other way.  In addition to recovering from downstream brokers, I think we are at least in principle capable of reconstructing alerts from the PPDB.

Author: I've revised to say that we should run a Kafka consumer which copies data into the alert database.

 1. Compute the alert packet payload, including generating a unique alert ID which should eventually appear in the PPDB.
 2. Write the alert packet to the object store, using the alert ID as a key.
 3. Publish the alert packet to the Kafka topic that serves data to community brokers.

In addition, after all observing is complete for a night, all the alerts that were succesfully published that night should be combined into a single Avro Object Container file.
The file should be stored on an archival filesystem.
The set of published alerts can be identified by consuming from the Kafka topic.

Reading data
------------

> To read individual alert data, users access the backing alert packets through the butler, which should wrap up the object storage and provide access by alert ID.

Reviewer: From our discussion [here](https://confluence.lsstcorp.org/display/~ebellm/Hybrid+Alerts+Discussion%3A+2021-03-25) I think we agreed that the access was through a datalink-type API query (`https://data.lsst.cloud/api/fullAlerts&alertID=xyzzy`, hypothetically).

Reviewer: Though I do note that Frossie suggested using the remote Butler for the MVP.

Author: I think of the Hybrid Alert system as a separate API layered on top of this one. We can present that datalink-type API separately, since it's easy to translate into the alert DB queries.

That Hybrid Alert API has other needs anyway, like rate limiting.

> To read bulk alert data, users can request the Avro Object Container files for particular nights.

Reviewer: How?  Through a different API endpoint?

Author: Hm, this is a good question, and one I don't have an answer for.

A (HTTP) API endpoint seems a little wrong, because I don't expect that a single gigantic TCP flow is the right way to send that data out. What sort of mechanisms do we use for bulk transfer elsewhere?

I guess as a fallback we could just do the HTTP endpoint...

Reviewer: Okay, maybe just flag that we need to identify the protocol here.

This satisfies each of the three use cases:
 - As a **historical record**: By writing to the object store first, we can be sure that all published alerts are recorded.
   In case of Kafka downtime, we may store _more_ alerts than were recorded, but this is acceptable.

   Archival files are available for bulk analysis of the historical record.
 - As a **queryable DB**: By querying the PPDB, users can search alerts by any of their fields or attributes, albeit with a one-day delay. Once they have alert IDs, they can get all underlying packets.
 - As a **resource for developers**: Object Container files provide bulk access.

Alert Identifier
----------------

We need an identifier which is unique across all alerts which can be used as the key for the object store.

> We can use ``diaSourceId`` for this purpose.

Reviewer: The current PPDB calls for a separate `alertId` separate from `diaSourceId`, to allow for the potential that the mapping is not 1-1.  I _think_ that's a good idea, but maybe you could convince me otherwise.  This document uses `alert_id` throughout, so we should at least be internally self-consistent here.

Author: I don't understand when the mapping wouldn't be 1-1. To me, those two labels are synonyms, so maybe my mental model is wrong.

Reviewer: I don't have a clear vision of this right now; most of the situations I can think of involve some kind of problem.  The best I can do is to say that I think each `diaSource` corresponds to _at most_ one alert (if we had classes of `diaSource` we didn't alert on because we knew they were junk, maybe?  Or if alert generation failed for some reason?), so a separate identifier helps answer the question of "where is the alert for this `diaSource`"--it doesn't exist!

I'm not totally convinced it's necessary, though.

Author: Okay. Well, if the PPDB already has an `alertId`, we can use that - I really don't have an opinion either way, just didn't realize that `alertId` exists and is separate from `diaSourceId`.

Limitations
===========

This design does not provide any sort of complex querying logic for data which has been stored since the last PPDB update.
Since the PPDB is updated daily, this means that the last 24 hours of data will not be indexed for complex queries.
This is acceptable, though, since the querying features of the alert database are not intended to support real-time online use cases.

Possible interaction with Alert Filtering Service
=================================================

> One possible design of an alert filtering service would be to publish alert packet IDs with a small batch of useful information about the alert.

Reviewer: I'd link to the Hybrid Alerts technote here.

To avoid a dependency upon a running Confluent Schema Registry for archive operations, we should store the schema document in the alert archive, indexed by its schema ID.

Since the schema ID is a 4-byte sequence, but object keys are ASCII text, we use a hex encoding of the schema ID.

The schema document that is stored should be a single Avro ``record`` which describes the alert packet.
Referenced subschemas should be transcluded into the document, and it should be stored in Avro's `Parsing Canonical Form`_ format.

.. _Confluent Wire Format: https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#wire-format
.. _Parsing Canonical Form: http://avro.apache.org/docs/current/spec.html#Parsing+Canonical+Form+for+Schemas

Schema updates
--------------

When a new version of the alert schema is released, the new schema should be written into the alert archive.
This can be done before any alerts are published with the new schema.

Writing data
------------

When the alert production pipeline has computed a new alert packet, it will write it to a Kafka topic, broadcasting it to brokers.

> We should implement and run a consumer of this Kafka topic which copies messages into the object store.

Reviewer 2: If it's accessible via Butler, these will have to be ingested/registered by the consumer.  I'm not sure the Butler Registry is going to handle the number of datasets and speed of publication very well.

The alternative would be to write new plugins for the Butler to special-case this datastore and dataset type.  I'm not sure what that gains us, aside from several additional layers of overhead for a service that we may eventually need to be low-latency.

Running as a consumer of the Kafka topic adds several seconds of latency.
This is acceptable because none of the primary use cases for the database require tight latency bounds.

Reading data
------------

> To read individual alert data, users access the backing alert packets through the butler, which should wrap up the object storage and provide access by alert ID.

Reviewer 2: Since alerts can only be looked up by a single key and have no associated data, the Butler seems like overkill, in addition to having scaling problems as mentioned above.  I would much prefer a simple, authenticated service that either proxies access to the object store or even better simply redirects to pre-signed URLs.

This satisfies each of the use cases:

>  - As a **historical record**: By writing to the object store first, we can be sure that all published alerts are recorded.

Reviewer 2: Not exactly "first" anymore, since it's a peer Kafka consumer.

>    In case of Kafka downtime, we may store _more_ alerts than were recorded, but this is acceptable.

Reviewer 2: This is also no longer true.

In addition, to facilitate bulk access, we may store single-file archives of all alerts published each night.

This system should be straightforward to build and administer, and it should be cost effective.
It will require about 120 terabytes of space in the object store per year.

Design Inputs
=============

Use cases
---------

Broadly speaking, the Alert Database will have two primary uses:

1. It will serve as an authoritative historical record, archiving all the alert data that the project has sent out through alert streams.
2. It will also permit science users to retrieve individual alert packets using an identifier that they got from somewhere else (for example, from a circular, or a query to the PPDB).

In addition, we aim to be flexible enough to support two possible additional use cases, without immediately committing to their development:

3. We may wish to provide bulk access to large piles of alert data (for example, several entire nights' alerts) for development and training of algorithms which process alert data.

> 4. We may wish to act as the low-level backend which supports an alert filtering service.

Reviewer: ```suggestion
4. We may wish it to serve as the low-level backend which supports an alert filtering service.
```

.. _queryable DB:

Use cases: queryable DB
^^^^^^^^^^^^^^^^^^^^^^^

The Alert Database needs to support "needle in a haystack" queries based on specific a single alert identifier.
In particular, ``get(alertId)`` ought to retrieve the single alert packet associated with a specific ``alertId``.

This should generally respond within a few seconds.

In general, we can expect recent data to be queried more often than old data for this purpose, but everything needs to be available.

Possible use case: bulk access
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Users who want to test algorithms, generate simulated streams, or evaluate filters will be interested in retrieving large batches of alert packets.
They will not need to retrieve these batches very often, and it's entirely acceptable for their requests to take a long time (even days) to fulfill.

We plan to see if there is suitable demand for this feature to justify adding it; supporting it might double our storage requirements.

Possible use case: alert filtering backend
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> DMTN-165 :cite:`DMTN-165` proposes an alert filtering service which consists of a stream of alert packet IDs and a small bit of associated data.

Reviewer: ```suggestion
DMTN-165 :cite:`DMTN-165` proposes a "hybrid" alert filtering service which consists of a stream of alert packet IDs and a small bit of associated data.
```

   But this would be more complex, and make writing more difficult, as writes need to append to existing data which would require coordination between writers.
   It would also make reading more complex; a separate index would need to be maintained which translates alert packet IDs into an identifier for the block containing the alert.
   In light of these complexities, this design sticks to a simpler structure.

An "object store" here is taken to mean an S3-like system.
It should it scale well to handle many terabytes of data.
The underlying object store should provide strong guarantees on data durability, either through redundancy or automated backups.
It should also support write throughput which can keep up with the alert stream volume, handling at least 2,000 writes per second [1]_, with each write averaging around 40 kilobytes.
Finally, it should support random read access with a median latency of under 2 seconds, permitting several hundred reads per second.

Writes to the object store are handled by a Kafka consumer which copies alert packets from the main Kafka topic into the alert database.

> Reads are servered with a lightweight HTTP service and accompanying client library which allow retrieval by alert ID of any packet.

Reviewer: Is "servered" spelled as you wanted?  I could see it either way!

Author: Oops, no, thanks.

> Reads are servered with a lightweight HTTP service and accompanying client library which allow retrieval by alert ID of any packet.

Author: ```suggestion
Reads are served with a lightweight HTTP service and accompanying client library which allow retrieval by alert ID of any packet.
```

To make that process faster, we could precompute bulk data files by adding another Kafka consumer process which builds hourly or nightly data batches, but this would come at the cost of duplicating storage.

Limitations
===========

No complex queries for last day of data
---------------------------------------

This design does not provide any sort of complex querying logic for data which has been stored since the last PPDB update.
Since the PPDB is updated daily, this means that the last 24 hours of data will not be indexed for complex queries.
This is acceptable, though, since the querying features of the alert database are not intended to support real-time online use cases.

Alerts are published before archival
------------------------------------

Alerts are published to brokers before they are archived, which minimizes latency to the brokers.
This introduces some risk of data loss.
If the archiving Kafka consumer fails or is misconfigured, we might broadcast alerts out to brokers without ever storing them in the alert database.

We have three fallbacks, however:

1. Kafka stores messages for a configurable length of time.

>    If the archivist recovers within the lifetime of messages in Kafka, we could replay historical alerts and write them into the object store.

Reviewer: ```suggestion
   If the archivist recovers within the lifetime of messages in Kafka (~one week), we could replay historical alerts and write them into the object store.
```
